drivers/wifi/nrf700x/osal/hw_if/hal/src/hal_fw_patch_loader.c

```diff
@@ -13,6 +13,95 @@
 #include "hal_fw_patch_loader.h"
 #include "hal_mem.h"
 
+/* To reduce HEAP maximum usage */
+#define MAX_PATCH_CHUNK_SIZE 8192
+#define ARRAY_SIZE(array) (sizeof(array) / sizeof((array)[0]))
+
+struct patch_contents {
+	const char *id_str;
+	const void *data;
+	unsigned int size;
+	unsigned int dest_addr;
+};
+
+/* In order to save RAM, divide the patch in to chunks download */
+static enum wifi_nrf_status hal_fw_patch_load(struct wifi_nrf_hal_dev_ctx *hal_dev_ctx,
+						enum RPU_PROC_TYPE rpu_proc,
+						const char *patch_id_str,
+						unsigned int dest_addr,
+						const void *fw_patch_data,
+						unsigned int fw_patch_size)
+{
+	enum wifi_nrf_status status = WIFI_NRF_STATUS_FAIL;
+	int last_chunk_size = fw_patch_size % MAX_PATCH_CHUNK_SIZE;
+	int num_chunks = fw_patch_size / MAX_PATCH_CHUNK_SIZE +
+					(last_chunk_size ? 1 : 0);
+
+	for (int chunk = 0; chunk < num_chunks; chunk++) {
+		unsigned char *patch_data_ram;
+		unsigned int patch_chunk_size =
+			((chunk == num_chunks - 1) ? last_chunk_size : MAX_PATCH_CHUNK_SIZE);
+		const void *src_patch_offset = (const char *)fw_patch_data +
+			chunk * MAX_PATCH_CHUNK_SIZE;
+		int dest_chunk_offset = dest_addr + chunk * MAX_PATCH_CHUNK_SIZE;
+
+		patch_data_ram = wifi_nrf_osal_mem_alloc(hal_dev_ctx->hpriv->opriv,
+									patch_chunk_size);
```

Comment:
Can the patch_data_ram (of MAX_PATCH_CHUNK_SIZE) be allocated and freed only once instead of allocing/freeing for every chunk?

Reply:
MAX_PATCH_CHUNK_SIZE refers to chunk, so, a bit confused by the statement. But we should be doing alloc per-chunk to minimise RAM usage.

Reply:
What I meant was reusing the same allocated memory to copy the chunks rather than allocating/freeing for every chunk.

Reply:
Offline discussion: As allocations are fast in Zephyr, the current approach actually saves both memory and time, so ,leaving it as is.